File: ios/StationToStation/Data/Exchange/ContactIdentity.swift
import CryptoKit
import Foundation
import Security

/// The device's durable **Contact**-facing identity (#265): one Secure Enclave keypair,
/// generated once and reused across every Exchange and every later LAN reconcile session.
/// The iOS twin of Android's `exchange/ContactIdentity.kt` and its `AndroidKeyStore` key.
///
/// This is what a **Friend**'s `publicKey` pins to, and what a LAN peer later proves
/// possession of via a signature instead of a fresh QR scan. Unlike the per-session TLS
/// certificate (`ContactCert`), it is meant to outlive everything: rotating it silently
/// unmakes every Contact relationship this device has.
///
/// ECDSA P-256, non-exportable, hardware-protected. Both platforms land on the same curve
/// without either having chosen it freely — the Secure Enclave only does P-256, and
/// `AndroidKeyStore` only does Ed25519 from API 33 — which is a coincidence worth relying
/// on and worth writing down.
///
/// **Not unit-tested**, for the same reason `AndroidKeyStoreCert` is not: the Secure
/// Enclave is real hardware that does not exist off-device. `signChallenge` and
/// `verifyChallenge` are where the signature math is pure and covered.
enum ContactIdentity {

    /// One keychain item, named once. Changing this string is a device-wide identity
    /// rotation, not a rename.
    private static let tag = Data("io.github.magnusencoded.stationtostation.contact-identity".utf8)

    /// The Exchange runs several sessions at once and each one wants the identity. Two
    /// threads racing `create()` would leave two keys under one tag and hand out
    /// whichever the keychain answered with — an identity that changes between sessions,
    /// which is exactly the failure this key exists to prevent.
    private static let lock = NSLock()
    private static var cached: SecKey?

    /// Base64 X.509 SubjectPublicKeyInfo — what goes on a `ProbeCard`, what is persisted
    /// on the far end's `Friend`, and what a challenge is later verified against.
    static func publicKeyBase64() -> String? {
        guard let key = key(),
              let publicKey = SecKeyCopyPublicKey(key),
              let x963 = SecKeyCopyExternalRepresentation(publicKey, nil) as Data?,
              // CryptoKit does the SubjectPublicKeyInfo wrapping, so the one place the
              // encoding could drift from Android's is a framework's problem, not ours.
              let parsed = try? P256.Signing.PublicKey(x963Representation: x963)
        else { return nil }
        return parsed.derRepresentation.base64EncodedString()
    }

    /// Signs `data` with the persisted identity — the proof a LAN peer checks against the
    /// public key already on their Friend record. Key material never leaves the enclave;
    /// only a reference to it is ever held here.
    ///
    /// `ecdsaSignatureMessageX962SHA256` is SHA-256 then ECDSA with a DER-encoded
    /// signature: byte for byte what Android's `"SHA256withECDSA"` produces.
    static func sign(_ data: Data) -> Data? {
        guard let key = key() else { return nil }
        return SecKeyCreateSignature(key, .ecdsaSignatureMessageX962SHA256,
                                     data as CFData, nil) as Data?
    }

    private static func key() -> SecKey? {
        lock.lock()
        defer { lock.unlock() }
        if let cached { return cached }
        let key = load() ?? create()
        cached = key
        return key
    }

    private static func load() -> SecKey? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassKey,
            kSecAttrApplicationTag as String: tag,
            kSecAttrKeyType as String: kSecAttrKeyTypeECSECPrimeRandom,
            kSecReturnRef as String: true,
        ]
        var item: CFTypeRef?
        guard SecItemCopyMatching(query as CFDictionary, &item) == errSecSuccess,
              let result = item, CFGetTypeID(result) == SecKeyGetTypeID()
        else { return nil }
        return (result as! SecKey)
    }

    /// The enclave first, a plain keychain key only if it refuses.
    ///
    /// The fallback is not a security compromise being waved through — it is the
    /// Simulator, which has no enclave, and without it every test run and every CI build
    /// would exercise a code path that simply returns nil. On a real phone the first
    /// branch is the one that runs.
    private static func create() -> SecKey? {
        let privateAttributes: [String: Any] = [
            kSecAttrIsPermanent as String: true,
            kSecAttrApplicationTag as String: tag,
        ]
        let attributes: [String: Any] = [
            kSecAttrKeyType as String: kSecAttrKeyTypeECSECPrimeRandom,
            kSecAttrKeySizeInBits as String: 256,
            kSecPrivateKeyAttrs as String: privateAttributes,
        ]

        // `.privateKeyUsage` alone: no biometry, no passcode prompt. The enclave protects
        // the key from extraction, and nothing here should ever interrupt someone to ask
        // permission for a sync they did not initiate.
        if let access = SecAccessControlCreateWithFlags(
            nil, kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly, .privateKeyUsage, nil
        ) {
            var enclave = attributes
            enclave[kSecAttrTokenID as String] = kSecAttrTokenIDSecureEnclave
            enclave[kSecPrivateKeyAttrs as String] = privateAttributes.merging(
                [kSecAttrAccessControl as String: access], uniquingKeysWith: { _, new in new }
            )
            if let key = SecKeyCreateRandomKey(enclave as CFDictionary, nil) { return key }
        }
        return SecKeyCreateRandomKey(attributes as CFDictionary, nil)
    }
}

File: android/app/src/main/java/io/github/magnusencoded/stationtostation/data/exchange/ContactIdentity.kt
package io.github.magnusencoded.stationtostation.data.exchange

import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyProperties
import java.security.KeyPairGenerator
import java.security.KeyStore
import java.security.PrivateKey
import java.security.spec.ECGenParameterSpec
import java.util.Base64

private const val ALIAS = "contact-identity"

/**
 * The device's durable Contact-facing identity (#257): one AndroidKeyStore keypair,
 * generated once and reused across every Exchange and every later LAN reconcile
 * session — unlike #142's ephemeral per-handover cert, this is the identity a Friend's
 * [io.github.magnusencoded.stationtostation.data.Friend.publicKey] pins to, and what a
 * LAN beacon later proves possession of via [signWithContactIdentity] instead of a
 * fresh QR scan.
 *
 * ECDSA P-256, not Ed25519: native Ed25519 in AndroidKeyStore needs API 33+, and this
 * app's minSdk is 26. Same call #142's AndroidKeyStoreCert.kt already made, for the
 * same reason.
 *
 * Not unit-tested, for the same reason as AndroidKeyStoreCert.kt: AndroidKeyStore is a
 * real, often hardware-backed provider that doesn't exist off-device. [signChallenge]
 * and [verifyChallenge] are where the signature math is pure and covered.
 */
private fun keyStore(): KeyStore = KeyStore.getInstance("AndroidKeyStore").apply { load(null) }

private fun ensureGenerated(keyStore: KeyStore) {
    if (keyStore.containsAlias(ALIAS)) return
    val spec = KeyGenParameterSpec.Builder(ALIAS, KeyProperties.PURPOSE_SIGN)
        // SHA-256 only, and it has to stay that way: this key is already on every device
        // that has ever made a Contact, and AndroidKeyStore keys are immutable — widening
        // this would mean a new key, and a new key is a new identity every existing
        // Contact would stop recognising. It signs challenges, never a TLS handshake:
        // TLS needs DIGEST_NONE, which is why a reconcile session mints its own
        // certificate (`AndroidKeyStoreCert.generateContactSessionIdentity`).
        .setDigests(KeyProperties.DIGEST_SHA256)
        .setAlgorithmParameterSpec(ECGenParameterSpec("secp256r1"))
        .build()
    KeyPairGenerator.getInstance(KeyProperties.KEY_ALGORITHM_EC, "AndroidKeyStore")
        .apply { initialize(spec) }
        .generateKeyPair()
}

/** Base64 X.509 SubjectPublicKeyInfo for the device's persisted identity — what goes on a Friend/ProbeCard. */
fun contactIdentityPublicKeyBase64(): String {
    val keyStore = keyStore()
    ensureGenerated(keyStore)
    return Base64.getEncoder().encodeToString(keyStore.getCertificate(ALIAS).publicKey.encoded)
}

/** The device's persisted identity key, for [mutualContactAuth] to sign with directly —
 * a reference into AndroidKeyStore, never the raw key material. */
fun contactIdentityPrivateKey(): PrivateKey {
    val keyStore = keyStore()
    ensureGenerated(keyStore)
    return keyStore.getKey(ALIAS, null) as PrivateKey
}

/** Signs [nonce] with the device's persisted identity — the proof a LAN peer checks against the public key already on their Friend record. */
fun signWithContactIdentity(nonce: ByteArray): ByteArray = signChallenge(nonce, contactIdentityPrivateKey())
